modules/library/metadata/src/main/java/org/geotools/factory/CategoryRegistry.java

```diff
@@ -0,0 +1,319 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2017, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.factory;
+
+import org.geotools.util.PartiallyOrderedSet;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.collectingAndThen;
+import static java.util.stream.Collectors.toMap;
+import static org.geotools.util.Utilities.ensureArgumentNonNull;
+import static org.geotools.util.Utilities.stream;
+
+/**
+ * The category registry holds multiple instances per category. Categories are
+ * {@link Class classes} and instances are also accessible by the class they implement.
+ * Note that instances have to implement/extend the category they are filed under.
+ *
+ * This class is {@code null}-intolerant and throws {@link IllegalArgumentException}s if
+ * an argument is {@code null}.
+ */
+class CategoryRegistry {
+
+	/**
+	 * Heterogeneous map, where `Class<T>` is mapped to `InstanceRegistry<T>` for any `T`.
+	 *
+	 * Code depends on this map never being mutated to avoid concurrent modification exceptions.
+	 */
+	private final Map<Class<?>, InstanceRegistry<?>> categories;
+
+	/**
+	 * Creates a new registry with the specified Registers the specified category.
+	 * If the same category is registered multiple times, all instances previously
+	 * registered for that category are lost.
+	 *
+	 * @param factoryRegistry The {@link FactoryRegistry} this registry belongs to.
+	 * @param categories The categories to register; must not be {@code null} but can contain {@code null}.
+	 */
+	public CategoryRegistry(final FactoryRegistry factoryRegistry, final Iterable<Class<?>> categories) {
+		ensureArgumentNonNull("factoryRegistry", factoryRegistry);
+		ensureArgumentNonNull("categories", categories);
+		// use an unmodifiable map to guarantee immutability
+		this.categories = stream(categories)
+				.collect(collectingAndThen(
+						toMap(
+								category -> category,
+								category -> new InstanceRegistry<>(factoryRegistry, category),
+								(firstCategory, secondCategories) -> secondCategories),
```

Comment:
`Collectors::toMap` enforces explicit handling of duplicate keys. The two-parameter variant throws an `IllegalStateException` if the same key is encountered twice. The variant I call here requires me to define what happens with the values of duplicate keys. Since both registries are empty, it does not matter which one I choose. (My naming is off: It should be `firstRegistry`/`secondRegistry`, not `firstCategory`/`secondCategory`.)